### DISCUSSION TITLE
uvicorn.run() not respecting the reload_excludes parameter?

### DISCUSSION THREAD

**Discussion by abceleung** [Kludex/uvicorn#1656] [Potential Issue] [answered]
I am using FastAPI with Python built-in logging module. 

Whenever the log file is updated, the program reloads as well, creating an infinite loop.

(I am using uvicorn.run() for debugging as suggested in the [FastAPI doc](https://fastapi.tiangolo.com/tutorial/debugging/))

Maybe I made a mistake somewhere? Here is my script:

```python
from fastapi import FastAPI
import uvicorn, logging
from logging.handlers import TimedRotatingFileHandler

logger = logging.getLogger()
logger.setLevel(logging.INFO)
logHandler = TimedRotatingFileHandler(
    "api.log",
    when="D",
    backupCount=10,
    encoding="utf-8",
)
logHandler.setFormatter(
    logging.Formatter(
        "%(asctime)s %(levelname)-8s %(name)-15s %(message)s",
        datefmt="%H:%M:%S",
    )
)
logger.addHandler(logHandler)

app = FastAPI()

@app.get("/")
async def root():
    return {"message": "Hello World"}
    
if __name__ == "__main__":
    uvicorn.run(
        "__main__:app",
        host="0.0.0.0",
        port=8000,
        reload=True,
        reload_excludes="*.log"
    )
```

Note: I have `watchfiles` installed.

**Comment by euri10** [comment]
how do you run uvicorn ? what exclude parameters are you passing ? what is the directory structure like, where the log file is located ?

**Reply by abceleung** [reply]
> how do you run uvicorn ? what exclude parameters are you passing ? what is the directory structure like, where the log file is located ?

> how do you run uvicorn ? what exclude parameters are you passing ? what is the directory structure like, where the log file is located ?

1. `python main.py`  <- `main.py` is the filename
2.  `reload_excludes="*.log"`, as shown in the script
3. As shown below (I am using Windows 10):
```
project
│   main.py
│   api.log (Generated by logger)   
│
└───.venv
      │   python libraries...

```

**Comment by euri10** [comment]
cant reproduce on linux though the api.log file is filled wuite rapidely with `10:30:06 INFO     watchfiles.main 1 change detected`, I seem to recall there was something with watchfiles on windows, maybe someone will remember

**Reply by abceleung** [reply]
> api.log file is filled wuite rapidely with `10:30:06 INFO watchfiles.main 1 change detected`

This is the infinite loop I am talking about

**Comment by euri10** [comment, marked as answer]
`reload_excludes` just tells uvicorn not to trigger a reload if  there is a change in the file, which is what is happening: api.log is changing and uvicorn does not reload itself, so everything works fine to me.

**Reply by abceleung** [reply]
I see. Is there something I can do with uvicorn to prevent this infinite logging behavior?

**Reply by euri10** [reply]
Yes, you can tell in your logging config to not log watchfiles messages or set a level that won't display info logs

**Reply by Bazzu85** [reply]
can you please share how to exclude this files from logging config?

**Reply by ridham-infocusp** [reply]
one way I have achieved infinite logging is by creating a filter and adding that to the root logger

class IgnoreChangeDetectedFilter(logging.Filter):
  def filter(self, record: logging.LogRecord):
    return '%d change%s detected: %s' != record.msg

root_logger = logging.getLogger()
root_logger.addFilter(IgnoreChangeDetectedFilter())

**Reply by obrua** [reply]
> one way I have achieved infinite logging is by creating a filter and adding that to the root logger
> 
> class IgnoreChangeDetectedFilter(logging.Filter): def filter(self, record: logging.LogRecord): return '%d change%s detected: %s' != record.msg
> 
> root_logger = logging.getLogger() root_logger.addFilter(IgnoreChangeDetectedFilter())

lookthis:
`
for name in logging.root.manager.loggerDict.keys():
    logging.getLogger(name).handlers = []
    if "watchfiles.main" in name:
        logging.getLogger(name).propagate = False
    else:
        logging.getLogger(name).propagate = True
  
  #logging.info(f"{logging.getLogger(name)}, {logging.getLogger(name).propagate}")
`

**Reply by socratescli** [reply]
The log is from a function in watchfiles

```
def _log_changes(changes: Set[FileChange]) -> None:
    if logger.isEnabledFor(logging.INFO):  # pragma: no branch
        count = len(changes)
        plural = '' if count == 1 else 's'
        if logger.isEnabledFor(logging.DEBUG):
            logger.debug('%d change%s detected: %s', count, plural, changes)
        else:
            logger.info('%d change%s detected', count, plural)
```

Setting the logging level for logger `watchfiles.main` could disable it
```
import logging
logger = logging.getLogger('watchfiles.main')
logger.setLevel(logging.WARNING)
```

**Reply by vavilon2142** [reply]
had the same issue, the problem is that when flag "reload=True" and your logs file in the same directory where app is, it constantly detects new logs and writes the log into file.
So the solution is to allocate logs file outside the app directory 
for instance: your directory for app is /app/main.app then logs should be outside the app folder. put them to /opt/logs for example
